### ISSUE TITLE
Concealing via extmarks

### ISSUE THREAD

**Issue by mrossinek** [neovim/neovim#15706]
### Feature already in Vim?

No

### Feature description

I recently raised a question over on nvim-treesitter about [treesitter-based concealing](https://github.com/nvim-treesitter/nvim-treesitter/discussions/1825#discussioncomment-1349544). I was redirected here as a feature like that would require a concealing method based on extmarks.
Is anything like that already planned or a WIP? I know about #9496 and #15351 which are both related to more advanced extmarks/virtual text but I don't think either of those cover this idea.

To make things a little bit more concrete, this feature request is about a mechanism to allow concealing of (e.g.) simple markup syntax when using treesitter highlighting. This would remove the need for a dual approach of TS-based highlighting in combination with the regex-based `syntax-region` definitions.

**Comment by mrossinek** [comment]
Actually I may have figured out a hacky (?) way to achieve this :thinking: 
In theory it's possible to replace the icons which one would want to conceal with a zero-width character like e.g. the [word joiner](https://en.wikipedia.org/wiki/Word_joiner). This works as expected but leads to some odd edge cases when editing the line in which such an extmark is set because there is an offset of the visible and actual text.
This can be fixed by deleting the extmark on the current cursor line and re-adding it once the cursor moved. This can obviously come with some serious performance penalty though...

So as a followup I would like to ask whether you think a new option on extmarks which hides them when the cursor is above their start position would be something worth adding directly into the core? I believe this will be better than deleting and re-adding on each `CursorMoved` or `CursorMovedI` event.,

**Comment by mrossinek** [comment]
Example screenshots:

Actual text:
![screenshot_1631959938](https://user-images.githubusercontent.com/21973473/133885345-de8812b8-eef9-425e-9305-d43cbbac6455.png)

Extmark on top using the word joiner (`U+2060`) instead of the `*` markup:
![screenshot_1631959951](https://user-images.githubusercontent.com/21973473/133885361-b9b6af15-c8ef-4653-ba9e-f3f05c2ff42c.png)

Notice that this leads to an edge case for the cursorcolumn at the end of the line so it is not ideal.

**Comment by XVilka** [comment]
Back in time I suggested this feature to Vim: https://github.com/vim/vim/issues/4687

**Comment by mrossinek** [comment]
Just a quick update on this: we recently merged some crude TS-based concealing into https://github.com/nvim-neorg/neorg using the word joiner "hack" which I described above. This does not work very consistently across systems unfortunately so we ended up disabling by default.
I am refactoring things right now and will probably resort to concealing via whitespace by default for the time being (simply to make things readable for users where the word joiner is not working as expected).

I simply write this update to mention a few observations, made by @vhyrro and myself:
- integrating a TS-based concealing approach into Neovim natively presumably will require a heavy refactoring of `screen.c` (maybe @bfredl has some insight into this, since you've been working on anti-conceal, virtual lines, and the like)
- a (in our opinion) nice approach for native TS-based concealing would involve some TS query files similar to this:
```
(node
  (nested_node_open) @conceal
  (node_content)
  (nested_node_close) @conceal
)
```
- :warning: something which may be a pre-requisite for all of the above is probably some optional extmark behavior which _hides extmarks under the cursor position_. I tried implementing this into neorg before but it was a rather significant performance penalty. I believe that a native solution to this could be much more efficient.

Is this last point something which is on anyone's radar? If not, should I open a separate feature request for this? I unfortunately do not have the time right now to be looking into something like this myself.
